sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/UnnecessaryImportsCheck.kt

```diff
@@ -0,0 +1,174 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.fir.builder.toBinaryName
+import org.jetbrains.kotlin.kdoc.psi.impl.KDocLink
+import org.jetbrains.kotlin.psi.KtArrayAccessExpression
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtImportDirective
+import org.jetbrains.kotlin.psi.KtImportList
+import org.jetbrains.kotlin.psi.KtOperationReferenceExpression
+import org.jetbrains.kotlin.psi.KtPackageDirective
+import org.jetbrains.kotlin.psi.KtReferenceExpression
+import org.jetbrains.kotlin.psi.KtSimpleNameExpression
+import org.jetbrains.kotlin.psi.KtUserType
+import org.jetbrains.kotlin.psi.KtVisitorVoid
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.psi.psiUtil.isDotSelector
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameOrNull
+import org.jetbrains.kotlin.resolve.descriptorUtil.getImportableDescriptor
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private const val MESSAGE_UNUSED = "Remove unused import."
+private const val MESSAGE_REDUNDANT = "Remove redundant import."
+
+@Rule(key = "S1128")
+class UnnecessaryImportsCheck : AbstractCheck() {
+
+    private fun Sequence<PsiElement>.flattenNodes(): Sequence<PsiElement> =
+        this + this.flatMap { it.children.asSequence().flattenNodes() }
+
+    override fun visitKtFile(file: KtFile, context: KotlinFileContext) {
+
+        val (references, arrayAccesses, kDocLinks) = collectData(file)
+        val bindingContext = context.bindingContext
+        val groupedReferences = references.groupBy { reference ->
+            reference.importableSimpleName()
+        }
+
+        val arrayAccessesImportsFilter by lazy {
+            getArrayAccessImportsFilter(arrayAccesses, bindingContext)
+        }
+
+        file.importDirectives.filter { imp ->
+            // 1. Filter out & report all imports that import from kotlin.* or the same package as our file
+            if (imp.isImportedImplicitlyAlready(file.packageDirective?.name)) {
+                imp.importedReference?.let { context.reportIssue(it, MESSAGE_REDUNDANT) }
+                false
+            } else true
+        }.groupBy {
+            it.importedName?.asString()
+        }.filterKeys { simpleName ->
```

Comment:
I'd suggest putting small one-line lambdas on the same line with the function invocation. so we'll have sth like:
```
 file.importDirectives
     .filter { ... }
     .groupBy { ... }
     .filterKeys { ... }
 ```

  And big lambdas could on the next lines or even become an extracted private methods